File: backends/python-service/app/services/cache.py
import json
from typing import Any, Optional, List, Tuple
import redis.asyncio as redis
from app.config import settings

# Global Redis client
_redis_client: Optional[redis.Redis] = None

async def get_redis_client() -> redis.Redis:
    """Get or create Redis client"""
    global _redis_client
    if _redis_client is None:
        _redis_client = redis.from_url(
            settings.REDIS_URL,
            encoding="utf-8",
            decode_responses=True
        )
    return _redis_client

class CacheService:
    def __init__(self, redis_client: redis.Redis):
        self.client = redis_client

    async def set(self, key: str, value: Any, expiration: int = 3600) -> bool:
        """Set a value in cache with expiration in seconds"""
        try:
            data = json.dumps(value, default=str)
            await self.client.setex(key, expiration, data)
            return True
        except Exception as e:
            print(f"Cache set error: {e}")
            return False

    async def get(self, key: str) -> Optional[Any]:
        """Get a value from cache"""
        try:
            data = await self.client.get(key)
            if data:
                return json.loads(data)
            return None
        except Exception as e:
            print(f"Cache get error: {e}")
            return None

    async def delete(self, key: str) -> bool:
        """Delete a key from cache"""
        try:
            await self.client.delete(key)
            return True
        except Exception as e:
            print(f"Cache delete error: {e}")
            return False

    async def exists(self, key: str) -> bool:
        """Check if key exists"""
        try:
            return bool(await self.client.exists(key))
        except Exception as e:
            print(f"Cache exists error: {e}")
            return False

    async def zadd(self, key: str, score: float, member: str) -> bool:
        """Add member to sorted set"""
        try:
            await self.client.zadd(key, {member: score})
            return True
        except Exception as e:
            print(f"Cache zadd error: {e}")
            return False

    async def zrevrange_withscores(
        self, key: str, start: int, stop: int
    ) -> List[Tuple[str, float]]:
        """Get range from sorted set in reverse order with scores"""
        try:
            result = await self.client.zrevrange(key, start, stop, withscores=True)
            return [(member, score) for member, score in result]
        except Exception as e:
            print(f"Cache zrevrange error: {e}")
            return []

    async def zrevrank(self, key: str, member: str) -> Optional[int]:
        """Get reverse rank of member in sorted set"""
        try:
            rank = await self.client.zrevrank(key, member)
            return rank
        except Exception as e:
            print(f"Cache zrevrank error: {e}")
            return None

    async def zscore(self, key: str, member: str) -> Optional[float]:
        """Get score of member in sorted set"""
        try:
            score = await self.client.zscore(key, member)
            return score
        except Exception as e:
            print(f"Cache zscore error: {e}")
            return None

    async def zcard(self, key: str) -> int:
        """Get cardinality of sorted set"""
        try:
            count = await self.client.zcard(key)
            return count or 0
        except Exception as e:
            print(f"Cache zcard error: {e}")
            return 0

    async def publish(self, channel: str, message: Any) -> bool:
        """Publish message to channel"""
        try:
            data = json.dumps(message, default=str)
            await self.client.publish(channel, data)
            return True
        except Exception as e:
            print(f"Cache publish error: {e}")
            return False

File: backends/go-service/internal/services/cache.go
package services

import (
	"context"
	"encoding/json"
	"time"

	"github.com/redis/go-redis/v9"
)

type CacheService struct {
	client *redis.Client
}

func NewCacheService(client *redis.Client) *CacheService {
	return &CacheService{
		client: client,
	}
}

// Set stores a value in cache with expiration
func (s *CacheService) Set(ctx context.Context, key string, value interface{}, expiration time.Duration) error {
	data, err := json.Marshal(value)
	if err != nil {
		return err
	}
	return s.client.Set(ctx, key, data, expiration).Err()
}

// Get retrieves a value from cache
func (s *CacheService) Get(ctx context.Context, key string, dest interface{}) error {
	data, err := s.client.Get(ctx, key).Bytes()
	if err != nil {
		return err
	}
	return json.Unmarshal(data, dest)
}

// Delete removes a key from cache
func (s *CacheService) Delete(ctx context.Context, key string) error {
	return s.client.Del(ctx, key).Err()
}

// Exists checks if a key exists
func (s *CacheService) Exists(ctx context.Context, key string) (bool, error) {
	count, err := s.client.Exists(ctx, key).Result()
	return count > 0, err
}

// Increment increments a counter
func (s *CacheService) Increment(ctx context.Context, key string) (int64, error) {
	return s.client.Incr(ctx, key).Result()
}

// ZAdd adds a member to a sorted set
func (s *CacheService) ZAdd(ctx context.Context, key string, score float64, member string) error {
	return s.client.ZAdd(ctx, key, redis.Z{
		Score:  score,
		Member: member,
	}).Err()
}

// ZRangeWithScores retrieves a range from sorted set with scores
func (s *CacheService) ZRangeWithScores(ctx context.Context, key string, start, stop int64) ([]redis.Z, error) {
	return s.client.ZRangeWithScores(ctx, key, start, stop).Result()
}

// ZRevRangeWithScores retrieves a range from sorted set in reverse order with scores
func (s *CacheService) ZRevRangeWithScores(ctx context.Context, key string, start, stop int64) ([]redis.Z, error) {
	return s.client.ZRevRangeWithScores(ctx, key, start, stop).Result()
}

// ZRank gets the rank of a member in sorted set
func (s *CacheService) ZRank(ctx context.Context, key, member string) (int64, error) {
	return s.client.ZRank(ctx, key, member).Result()
}

// ZRevRank gets the reverse rank of a member in sorted set
func (s *CacheService) ZRevRank(ctx context.Context, key, member string) (int64, error) {
	return s.client.ZRevRank(ctx, key, member).Result()
}

// ZScore gets the score of a member in sorted set
func (s *CacheService) ZScore(ctx context.Context, key, member string) (float64, error) {
	return s.client.ZScore(ctx, key, member).Result()
}

// Publish publishes a message to a channel
func (s *CacheService) Publish(ctx context.Context, channel string, message interface{}) error {
	data, err := json.Marshal(message)
	if err != nil {
		return err
	}
	return s.client.Publish(ctx, channel, data).Err()
}

// Subscribe subscribes to a channel
func (s *CacheService) Subscribe(ctx context.Context, channel string) *redis.PubSub {
	return s.client.Subscribe(ctx, channel)
}
